src/ReverseProxy/Service/HealthChecks/DestinationHealthUpdater.cs

```diff
@@ -0,0 +1,137 @@
+            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+        public void SetActive(ClusterInfo cluster, IEnumerable<NewActiveDestinationHealth> newHealthPairs)
+        {
+            var changed = false;
+            foreach (var newHealthPair in newHealthPairs)
+            {
+                var destination = newHealthPair.Destination;
+                var newHealth = newHealthPair.NewActiveHealth;
+
+                var healthState = destination.Health;
+                if (newHealth != healthState.Active)
+                {
+                    healthState.Active = newHealth;
+                    changed = true;
+                    if (newHealth == DestinationHealth.Unhealthy)
+                    {
+                        Log.ActiveDestinationHealthStateIsSetToUnhealthy(_logger, destination.DestinationId, cluster.ClusterId);
+                    }
+                    else
+                    {
+                        Log.ActiveDestinationHealthStateIsSet(_logger, destination.DestinationId, cluster.ClusterId, newHealth);
+                    }
+                }
+            }
+
+            if (changed)
+            {
+                cluster.UpdateDynamicState();
+            }
+        }
+
+        public Task SetPassiveAsync(ClusterInfo cluster, DestinationInfo destination, DestinationHealth newHealth, TimeSpan reactivationPeriod)
+        {
+            var healthState = destination.Health;
+            if (newHealth != healthState.Passive)
+            {
+                healthState.Passive = newHealth;
+                ScheduleReactivation(cluster, destination, newHealth, reactivationPeriod);
+                return Task.Factory.StartNew(c => ((ClusterInfo)c).UpdateDynamicState(), cluster, TaskCreationOptions.RunContinuationsAsynchronously);
+            }
+            return Task.CompletedTask;
+        }
+
+        private void ScheduleReactivation(ClusterInfo cluster, DestinationInfo destination, DestinationHealth newHealth, TimeSpan reactivationPeriod)
+        {
+            if (newHealth == DestinationHealth.Unhealthy)
+            {
+                _scheduler.ScheduleEntity((cluster, destination), reactivationPeriod);
+                Log.UnhealthyDestinationIsScheduledForReactivation(_logger, destination.DestinationId, reactivationPeriod);
+            }
+        }
+
+        public void Dispose()
+        {
+            _scheduler.Dispose();
+        }
+
+        private Task Reactivate(ClusterInfo cluster, DestinationInfo destination)
+        {
+            var healthState = destination.Health;
+            if (healthState.Passive == DestinationHealth.Unhealthy)
+            {
+                healthState.Passive = DestinationHealth.Unknown;
+                cluster.UpdateDynamicState();
+                Log.PassiveDestinationHealthResetToUnkownState(_logger, destination.DestinationId);
```

Comment:
Log before continuing with the update, it's non-trivial.
```suggestion
                healthState.Passive = DestinationHealth.Unknown;
                Log.PassiveDestinationHealthResetToUnkownState(_logger, destination.DestinationId);
                cluster.UpdateDynamicState();
```